packages/logging/src/logging-and-telemetry.ts

```diff
@@ -0,0 +1,884 @@
+import redactInfo from 'mongodb-redact';
+import { redactURICredentials } from '@mongosh/history';
+import type {
+  MongoshBus,
+  ApiEventWithArguments,
+  ApiEvent,
+  UseEvent,
+  EvaluateInputEvent,
+  ShowEvent,
+  ConnectEvent,
+  ScriptLoadFileEvent,
+  StartLoadingCliScriptsEvent,
+  StartMongoshReplEvent,
+  GlobalConfigFileLoadEvent,
+  CryptLibrarySkipEvent,
+  CryptLibraryFoundEvent,
+  SnippetsCommandEvent,
+  SnippetsErrorEvent,
+  SnippetsFetchIndexErrorEvent,
+  SnippetsFetchIndexEvent,
+  SnippetsLoadedEvent,
+  SnippetsLoadSnippetEvent,
+  SnippetsNpmDownloadActiveEvent,
+  SnippetsNpmDownloadFailedEvent,
+  SnippetsNpmLookupEvent,
+  SnippetsRunNpmEvent,
+  SnippetsTransformErrorEvent,
+  EditorRunEditCommandEvent,
+  EditorReadVscodeExtensionsDoneEvent,
+  EditorReadVscodeExtensionsFailedEvent,
+  FetchingUpdateMetadataEvent,
+  FetchingUpdateMetadataCompleteEvent,
+  SessionStartedEvent,
+  MongoshBusEventsMap,
+  WriteCustomLogEvent,
+} from '@mongosh/types';
+import { inspect } from 'util';
+import { MongoLogWriter } from 'mongodb-log-writer';
+import { mongoLogId } from 'mongodb-log-writer';
+import type {
+  AnalyticsIdentifyMessage,
+  MongoshAnalytics,
+  MongoshAnalyticsIdentity,
+} from './analytics-helpers';
+import type { ConnectEventMap } from '@mongodb-js/devtools-connect';
+import { hookLogger } from '@mongodb-js/devtools-connect';
+import { MultiSet, toSnakeCase } from './helpers';
+import { Writable } from 'stream';
+
+export class MongoshLoggingAndTelemetry {
+  private log: MongoLogWriter | null;
+
+  private readonly trackingProperties: {
+    mongosh_version: string;
+    session_id: string;
+  };
+  private readonly userTraits: AnalyticsIdentifyMessage['traits'] & {
+    [key: string]: unknown;
+  };
+
+  private pendingLogEvents: CallableFunction[] = [];
+
+  private telemetryAnonymousId: string | undefined;
+  private userId: string | undefined;
+  private isDummyLog = true;
+  private isSetup = false;
+  private hookedExternalListenersLogId: string | undefined = undefined;
```

Comment:
I feel like this class's state space exploded a bit now, which makes this pretty hard to review and probably also hard to work with in the future

e.g. it's hard to tell whether `isDummyLog = false, isSetup = true, log = null, pendingLogEvents = [...], hookedExternalListenersLogId = 'foo'` is a valid state for this class or not and what that means specifically

I'll try to get more specific at individual points here but I think we might want to rethink this and look for opportunities to make this simpler